web/shared/src/typeahead.ts

```diff
@@ -48,14 +48,41 @@ export function remove_diacritics(s: string): string {
     return s.normalize("NFKD").replace(unicode_marks, "");
 }
 
+function upper_bound_lexicographically(prefix: string, words: string[]): number {
+    let left = 0;
+    let right = words.length;
+    let found = false;
+    while (left < right) {
+        const mid = Math.floor((left + right) / 2);
+        if (words[mid].startsWith(prefix)) {
+            left = mid + 1;
+            found = true;
+        } else if (words[mid] < prefix) {
+            left = mid + 1;
+        } else {
+            right = mid;
+        }
+    }
+    if (found) {
+        return left;
+    }
+    return 0;
```

Comment:
Hmm. What if `left` is 0? Can that happen? If so, is that a case that a caller ought to be able to distinguish from this `return 0;` case?

Consider returning `null` when your meaning is "there's nothing here", rather than a value like 0 that resembles the "yes, there is something here, here's what it is" values.

Reply:
Since the line `left = mid + 1` immediately precedes the only place we mark `found` as true, and `mid` is always >= 0, there can never be a situation where `found` is true _and_ `left` is 0. 

Generally speaking, the return value of the function is meant to be the index of the next string after the last prefix match of `prefix`. Hence we subtract 1 in the calling function to get that match. So a return value of 0 indicates that there is no previous element that would be the match... Hope that makes sense?

I'll add a comment explaining this among other things!